Makefile

```diff
@@ -92,133 +93,147 @@ deploy-live: prod
-test-hub: client hub-node-modules
-	echo coming soon!
+test-hub: hub database ethprovider
+	bash ops/test-hub.sh
 
-test-client: client hub-node-modules
-	echo coming soon!
+test-client: client
+	bash ops/test-client.sh
 
-test-integration: dev
-	echo coming soon!
+test-e2e: root-node-modules prod
+	npm stop
+	npm run prod
+	./node_modules/.bin/cypress run
+	npm stop
 
+########################################
 # Begin Real Rules
 
 # Proxy
 
 proxy-prod: wallet-prod $(shell find $(proxy) $(find_options))
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	docker build --file $(proxy)/prod.dockerfile --tag $(project)_proxy:latest .
 	$(log_finish) && touch build/proxy-prod
 
 proxy: $(shell find $(proxy) $(find_options))
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	docker build --file $(proxy)/dev.dockerfile --tag $(project)_proxy:dev .
 	$(log_finish) && touch build/proxy
 
 # Wallet
 
 wallet-prod: wallet-node-modules $(shell find $(wallet)/src $(find_options))
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	$(docker_run_in_wallet) "rm -f .env && cp ops/prod.env .env"
 	$(docker_run_in_wallet) "npm run build"
 	$(log_finish) && touch build/wallet-prod
 
 wallet: wallet-node-modules $(shell find $(wallet)/src $(find_options))
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	docker build --file $(wallet)/ops/dev.dockerfile --tag $(project)_wallet:dev $(wallet)
 	$(log_finish) && touch build/wallet
 
 wallet-node-modules: builder client $(wallet)/package.json
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	$(docker_run_in_wallet) "rm -rf node_modules/connext"
-	$(docker_run_in_wallet) "$(install)"
+	$(docker_run_in_wallet) "$(install)" > /dev/null
 	$(docker_run_in_wallet) "rm -rf node_modules/connext"
 	$(docker_run_in_wallet) "ln -s ../../client node_modules/connext"
+	$(docker_run_in_wallet) "cd ../client && $(install)" > /dev/null
 	$(log_finish) && touch build/wallet-node-modules
 
 # Hub
 
 hub-prod: hub-js
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	docker build --file $(hub)/ops/prod.dockerfile --tag $(project)_hub:latest .
 	$(log_finish) && touch build/hub-prod
 
 hub: hub-js $(hub)/ops/dev.entry.sh
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	docker build --file $(hub)/ops/dev.dockerfile --tag $(project)_hub:dev $(hub)
 	$(log_finish) && touch build/hub
 
 hub-js: hub-node-modules $(shell find $(hub) $(find_options))
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	$(docker_run_in_hub) "./node_modules/.bin/tsc -p tsconfig.json"
 	$(log_finish) && touch build/hub-js
 
 hub-node-modules: builder client $(hub)/package.json
-	$(log_start)
+	$(log_start) && echo "prereqs: $<"
 	$(docker_run_in_hub) "rm -rf node_modules/connext"
-	$(docker_run_in_hub) "$(install)"
+	$(docker_run_in_hub) "$(install)" > /dev/null
 	$(docker_run_in_hub) "rm -rf node_modules/connext"
 	$(docker_run_in_hub) "ln -s ../../client node_modules/connext"
+	$(docker_run_in_hub) "cd ../client && $(install)" > /dev/null
```

Comment:
Stupid npm aggressively tries to dedup packages & this often results in packages getting removed from the client's node_modules. An extra installation in the client fixes a lot of the missing dependency bugs we've been seeing.